### ISSUE TITLE
🐛 [BUG] - non_overlapping workflow creates invalid OTUs when data is a mix of overlapping and non-overlapping reads

### ISSUE THREAD

**Issue by marchoeppner** [bio-raum/FooDMe2#87]
### Description

Not a huge surprise, but for our fungal test data, some of the inserts are smaller and some are larger than the read length (we discussed this). 

This also has some consequences for the merging of reads, where --non_overlapping currently triggers the "glueing" of reads via normal joining rather than merging. 

Just looking at this manually, we do get reads that are joined but should be merged, resulting in them being filtered out during blast. 

```
GAAATGCGATAAGTAATGTGAATTGCAGAATTCAGTGAATCATCGAATCTTTGAACGCATCTTGCGCTCCTTGGTATTCCGAGGAGCATGCCTGTTTGAGTGTCATTATATTCTCAACTCTCCAATACTTTGTTGTAAAGGAGAGCTTGGATTGTGGAGGTTTGCTGGCTCCTTACTTGGGGTCAGCTCCTCTGAAATGCATTAGCGGAATCGTCTGCGATCTGCCACAAG
                                                                                                             NNNNNNNNNNCTCCAATACTTTGTTGTAAAGGAGAGCTTGGATTGTGGAGGTTTGCTGGCTCCTTACTTGGGGTCAGCTCCTCTGAAATGCATTAGCGGAATCGTCTGCGATCTGCCACAAGTGTGATAACTTATCTACACTGGCGAGGGGATTGCTTTCTGATGTTCAGCTTCTAATCGTCTAAGGACAATTTCTTGAATGCTTGACCTCAAATCAGGTAGGACTACCCGCTGA
```
as one example. 

How to deal with this?

In vsearch, the tool ` --fastq_mergepairs` is able to save all pairs that were not merged into a separate set of files, so one could potentially use these then for joining and combine merged and joined reads into one set of fastq files. 

Thoughts?

It's super annoying ;)

### Nextflow version

24.04.4

### Pipeline version

1.0.0

### Reproduction steps

```bash
Any dataset with overlapping and non-overlapping pe reads.
```

### Deployment

Singularity

### OS

Linux

**Comment by gregdenay** [comment]
For me this is bad experimental designon the user's side. 
We could add some workaround like you propose but that's getting pretty complicated and I#m not sure the use cases actually justify it. It gets kind of hard to say if merging failed because of low quality or because of no overlap. Probably should just use 2x250bp sequencing instead.

I had already added a mention of this in the doc of  the `non_ovelrapping` parameter,

**Comment by marchoeppner** [comment]
That is obviously another way to look at it; agreed on my solution risking to let reads through that failed to merge based on quality. On the other hand, if we flat-out join all the reads, the issue is also there. So I am thinking that doing it as per my idea would at least remove some of the incorrectly joined reads?

Short-reads for ITS is probably problematic any way you look at it...


```
"""

        vsearch --fastq_mergepairs $fwd --reverse $rev \
        --fastqout_notmerged_fwd forward_not_merged.fastq \
        --fastqout_notmerged_rev reverse_not_merged.fastq \
        --fastqout pe.fastq \
        --fastq_eeout \
        --relabel ${meta.sample_id}. \
        --sample ${meta.sample_id} \
        $args

        vsearch --fastq_join forward_not_merged.fastq --reverse reverse_not_merged.fastq \
        --fastqout join.fastq \
        --threads ${task.cpus} \
        --relabel ${meta.sample_id}. \


        cat pe.fastq join.fastq >> $merged
        rm pe.fastq join.fastq
```

**Comment by marchoeppner** [comment]
Ok, as discussed, I re-implemented the joining strategy, initially added in 7fdaba3

This works as follows:

Step 1: Merge all the reads as we normally would
Step 2: Take all the reads that could not be merge, any join them with an N space
Step 3: Combine fastq files from Step1+2 and proceed as normal

This runs through on our ITS test data, but the results still look like rubbish for most samples. But doing some manual blasts, my best guess is that the databases are to blame. I cannot test core_nt on my local dev machine, but I am guessing results there will be a lot better. 

Let me know if you agree with the solution.

**Comment by marchoeppner** [comment]
Right, so what I am noticing is that after the clustering in vsearch, the NNNNs are gone. So something there seems to be kicking these reads out. They are there going in at least...but after clustering, gone. I would not think that the joined reads should cluster with merged reads at all, unless Vsearch deliberately ignores Ns. Hm.

**Comment by gregdenay** [comment]
Are the reads gone or clustered with other (I assume then shorter) sequences?

There is quite a lot happening under the hood in VSearch clustering. Some things I can think of checking tohe doc are:
- disabling dust masking
```
--qmask none|dust|soft
Mask regions in sequences using the dust or the soft methods, or do not mask (none).
Warning, when using soft masking, clustering becomes case sensitive. The default is to
mask using dust.
```

- change ID calculation, although the available options only seem to act on terminal gaps:
```
--iddef 0|1|2|3|4
Change the pairwise identity definition used in --id. Values accepted are:
0. CD-HIT definition: (matching columns) / (shortest sequence length).
1. edit distance: (matching columns) / (alignment length).
2. edit distance excluding terminal gaps (same as --id).
3. Marine Biological Lab definition counting each gap opening (internal or
terminal) as a single mismatch, whether or not the gap was extended: 1.0
- [(mismatches + gap openings)/(longest sequence length)]
4. BLAST definition, equivalent to --iddef 1 in a context of global pairwise
alignment.
```

Should use th --uc option to ouptu the clustering results and be able to understand what's happening

**Comment by gregdenay** [comment]
I'm looking at this for DADA2 and it's a real mess.

What do we do with reads that overlap but the overlap is too short to join? Normally there is not enough support for mergign, but joining leads to a short repeated sequence that does not exist.

**Comment by gregdenay** [comment]
This behaviour is now also implemented for the DADA2 worfklow. Since DADA2 actually clusters the reads before merging here is what happens at the Denoising/merging step:

- Denoising of individual (filtered) reads using the error-models
- Merge reads and report all merging results (also fails)
- Concatenate all reads
- Use the merge status (pass or fail) to pick the sequences from either the merged (pass) or concatenated (fails) dataframes
- make the seqtable with the hybrid dataframe

**Comment by gregdenay** [comment]
We should probably test a bit more before closing

**Comment by gregdenay** [comment]
Closing this for now, as it is part of #99 

We should open fresh issues if needed once we have feedback on the feature
